tempesta_fw/sched.c

```diff
@@ -39,7 +40,138 @@
 static LIST_HEAD(sched_list);
 static DEFINE_SPINLOCK(sched_lock);
 
-/*
+/**
+ * Schedule connection to either to main @main_sg server group or to backup
+ * @backup_sg if scheduling to main group failed.
+ */
+static inline TfwConnection *
+sched_sg_conn(TfwMsg *msg, TfwSrvGroup *main_sg, TfwSrvGroup *backup_sg)
+{
+	TfwConnection *conn;
+
+	TFW_DBG2("sched %s: schedule from group: '%s'\n", main_sg->sched->name,
+		 main_sg->name);
+	conn = main_sg->sched->sched_sg_conn(msg, main_sg);
+
+	if (unlikely(!conn && backup_sg)) {
+		TFW_DBG("sched %s: the main group is offline, schedule from "
+			"backup group '%s'\n", backup_sg->sched->name,
+			backup_sg->name);
+		conn = backup_sg->sched->sched_sg_conn(msg, backup_sg);
+	}
+
+	if (unlikely(!conn))
+		TFW_DBG2("sched: Unable to select server from main and buckup "
+			 "group\n");
+
+	return conn;
+}
+
+/**
+ * Try to reuse last connection to @sg server group saved in http session.
+ * Fallback to scheduling from server group for newly setting connections to
+ * @sg or if sticky sessions failovering is enabled.
+ *
+ * @return TfwConnection *, NULL, or -1 if message cannot be scheduled and must
+ * be dropped.
+ */
+static TfwConnection *
+sched_conn_sticky(TfwMsg *msg, TfwSrvGroup *sg)
+{
+	TfwHttpReq *req = (TfwHttpReq *)msg;
+	TfwConnection *conn;
+
+	conn = tfw_http_sess_get_conn(req, sg);
+	if (conn) {
+		TfwServer *srv = (TfwServer *)conn->peer;
+
+		if (tfw_connection_get_if_nfo(conn))
+			return conn;
+
+		/*
+		 * @conn->peer->sg may differ from @sg if connection was
+		 * scheduled to backup group last time.
+		*/
+		if ((conn = srv->sg->sched->sched_srv_conn(msg, srv)))
+			return conn;
+
+		if (!tfw_cfg_sticky_sessions_failover) {
+			return ERR_PTR(-EINVAL);
+		}
+		else {
+			char addr_str[TFW_ADDR_STR_BUF_SIZE] = { 0 };
+
+			tfw_addr_ntop(&srv->addr, addr_str, sizeof(addr_str));
+			TFW_WARN("sched: Unable to reschedule request to the "
+				 "same server %s, schedule from server group\n",
+				 addr_str);
```

Comment:
No, that's not quite what I meant. I meant to be a bit more specific about which scheduler and group *were* used for scheduling to the server, not which group will be used.